# ch02 파이썬 웹 표준 라이브러리

### 예제 2-1 urlopen() 함수

In [1]:
from urllib2 import urlopen

In [7]:
f = urlopen('http://www.example.com')

In [8]:
print f.read(500)

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin:


## 예제 2-2 urlopen() 함수 - POST 방식 요청

In [9]:
from urllib2 import urlopen

In [10]:
data = 'query=python'

In [11]:
f = urlopen('http://www.example.com', data)

In [12]:
print f.read(300)

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background


### 예제 2-3 urlopen() 함수 - Request 클래스로 요청 헤더 지정

In [13]:
import urllib2
req = urllib2.Request('http://www.example.com')
req.add_header('Content-Type', 'text/plain')
req.add_data('query=python')
f = urllib2.urlopen(req)
print f.read(300)

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background


### 2.2.3 urllib2 모듈 예제

In [14]:
from urllib2 import urlopen
from HTMLParser import HTMLParser

class ImageParser(HTMLParser):
    def handle_starttag(self, tag, attrs):
        if tag != 'img':
            return
        if not hasattr(self, 'result'):
            self.result = []
        for name, value in attrs:
            if name == 'src':
                self.result.append(value)
def parseImage(data):
    parser = ImageParser()
    parser.feed(data)
    dataSet = set(x for x in parser.result)
    print '\n'.join(sorted(dataSet))

def main():
    url = 'http://www.google.co.kr'
    
    f = urlopen(url)
    charset = f.info().getparam('charset')
    data = f.read().decode(charset)
    f.close()
    
    print '\n>>>>>>> Fetch Images from', url
    parseImage(data)
    
main()


>>>>>>> Fetch Images from http://www.google.co.kr
/images/icons/product/chrome-48.png
/logos/doodles/2015/155th-anniversary-of-the-pony-express-5959391580782592.2-hp.jpg
/textinputassistant/tia.png


### 2.2.4 httplib 모듈

### 예제 2-8 httplib 모듈 사용 - GET 방식 요청

In [15]:
import httplib

In [28]:
conn = httplib.HTTPConnection('www.example.com')

In [29]:
conn.request('GET', '/index.html')

In [30]:
r1 = conn.getresponse()

In [31]:
r1.status, r1.reason

(200, 'OK')

In [32]:
data1 = r1.read()

In [33]:
print data1

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 50px;
        background-color: #fff;
        border-radius: 1em;
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        body {
            background-color: #fff;
        }
        div {
            width: auto;
            margin: 0 auto;
            border-radius: 0;
            padding: 1em;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This doma

In [34]:
conn.request('GET', '/parrot.spam')

In [35]:
r2 = conn.getresponse()

In [36]:
print r2.status, r2.reason

404 Not Found


In [37]:
data2 = r2.read()
print data2

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 50px;
        background-color: #fff;
        border-radius: 1em;
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        body {
            background-color: #fff;
        }
        div {
            width: auto;
            margin: 0 auto;
            border-radius: 0;
            padding: 1em;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This doma

In [38]:
conn.close()

### 예제 2-9 httplib 모듈 사용 - HEAD 방식 요청

In [39]:
import httplib
conn = httplib.HTTPConnection('www.example.com')
conn.request('HEAD', '/index.html')
res = conn.getresponse()
print res.status, res.reason

200 OK


In [40]:
data = res.read()
print len(data)

0


In [41]:
data == ''

True

### 예제 2-10 httplib 모듈 사용 - POST 방식 요청

In [42]:
import httplib, urllib

In [43]:
params = urllib.urlencode({'@number': 12524, 
                           '@type': 'issue', 
                           '@action': 'show'})
headers = {'Content-type': 'application/x-www-form-urlencoded',
           'Accept': 'text/plain'}

In [44]:
conn = httplib.HTTPConnection('bugs.python.org')

In [45]:
conn.request('POST', '', params, headers)

In [46]:
response = conn.getresponse()

In [47]:
print response.status, response.reason

302 Found


In [48]:
data = response.read()

In [49]:
print data

Redirecting to <a href="http://bugs.python.org/issue12524">http://bugs.python.org/issue12524</a>


In [50]:
conn.close()

### 예제 2-11 httplib 모듈 사용 - PUT 방식 요청

In [51]:
import httplib
BODY = '***filecontents***'
conn = httplib.HTTPConnection('localhost', 8888)
conn.request('PUT', '/file', BODY)
response = conn.getresponse()
print response.status, response.reason

404 Not Found


### 2.2.5 httplib 모듈 예제

In [52]:
import httplib
from urlparse import urljoin, urlunparse
from urllib import urlretrieve
from HTMLParser import HTMLParser
import os

class ImageParser(HTMLParser):
    def handle_starttag(self, tag, attrs):
        if tag != 'img':
            return
        if not hasattr(self, 'result'):
            self.result = []
        for name, value in attrs:
            if name == 'src':
                self.result.append(value)
                
def downloadImage(srcUrl, data):
    if not os.path.exists('DOWNLOAD'):
        os.makedirs('DOWNLOAD')
    
    parser = ImageParser()
    parser.feed(data)
    resultSet = set(x for x in parser.result)
    
    for x in sorted(resultSet):
        src = urljoin(srcUrl, x)
        basename = os.path.basename(src)
        targetFile = os.path.join('DOWNLOAD', basename)
        
        print 'Downloading...', src
        urlretrieve(src, targetFile)
        
def main():
    host = 'www.google.co.kr'
    conn = httplib.HTTPConnection(host)
    conn.request('GET', '')
    resp = conn.getresponse()
    
    charset = resp.msg.getparam('charset')
    data = resp.read().decode(charset)
    conn.close()
    
    print '\n>>>>>>> Download Images from', host
    url = urlunparse(('http', host, '', '', '', ''))
    downloadImage(url, data)
    
main()


>>>>>>> Download Images from www.google.co.kr
Downloading... http://www.google.co.kr/images/icons/product/chrome-48.png
Downloading... http://www.google.co.kr/logos/doodles/2015/155th-anniversary-of-the-pony-express-5959391580782592.2-hp.jpg
Downloading... http://www.google.co.kr/textinputassistant/tia.png


In [54]:
!ls -l DOWNLOAD/

total 160
-rw-r--r--  1 re4lfl0w  staff  72178  4 14 06:12 155th-anniversary-of-the-pony-express-5959391580782592.2-hp.jpg
-rw-r--r--  1 re4lfl0w  staff   1834  4 14 06:12 chrome-48.png
-rw-r--r--  1 re4lfl0w  staff    387  4 14 06:12 tia.png


In [55]:
!open DOWNLOAD/chrome-48.png

In [58]:
!open DOWNLOAD/tia.png

In [57]:
!open DOWNLOAD/155th-anniversary-of-the-pony-express-5959391580782592.2-hp.jpg

## 2.3 웹 서버 라이브러리

In [60]:
from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler

class MyHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.wfile.write('Hello World')
        
if __name__ == '__main__':
    server = HTTPServer(('', 8000), MyHandler)
    print 'Started WebServer on port 8888...'
    print 'Press ^C to quit WebServer'
    server.serve_forever()

Started WebServer on port 8888...
Press ^C to quit WebServer


KeyboardInterrupt: 

#### 표 2-4 웹 서버용 파이썬 라이브러리 모듈

모듈명 | 모듈에서 정의하고 있는 내용 | 처리 기능
--- | --- | ---
BaseHTTPServer | - 기반 서버 클래스용으로, HTTPServer 정의<br/>- 핸들러 클래스용으로, BaseHTTPRequestHandler 정의<br/>- 테스트용 웹 서버를 실행하는 함수, test() 정의 | 기반클래스로, HTTP 프로토콜 처리
SimpleHTTPServer | - 기반 서버 클래스인 HTTPServer를 임포트하여 사용<br/>- 핸들러 클래스용으로, SimpleHTTPRequestHandler 정의<br/>- 테스트용 웹 서버를 실행하는 함수, test() 정의 | GET과 HEAD 메소드 처리 가능
CGIHTTPServer | - 기반 서버 클래스인 HTTPServer를 임포트하여 사용<br/>- 핸들러 클래스용으로, CGIHTTPRequestHandler 정의<br/>- 테스트용 웹 서버를 실행하는 함수, test() 정의 | POST와 CGI 처리 가능

### 2.3.2 BaseHTTPServer 모듈

- 직접 MyHandler 라는 자신의 핸들러를 코딩하였음
- do_GET() 및 do_HEAD() 메소드가 미리 정의되어 있음
- POST 등 그 이외의 HTTP 메소드는 처리할 수 없음

In [62]:
!python -m SimpleHTTPServer 8001

Serving HTTP on 0.0.0.0 port 8001 ...
127.0.0.1 - - [15/Apr/2015 23:20:55] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [15/Apr/2015 23:20:56] code 404, message File not found
127.0.0.1 - - [15/Apr/2015 23:20:56] "GET /favicon.ico HTTP/1.1" 404 -
^CTraceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SimpleHTTPServer.py", line 224, in <module>
    test()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SimpleHTTPServer.py", line 220, in test
    BaseHTTPServer.test(HandlerClass, ServerClass)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/BaseHTTPServer.py", line 599, in tes

In [63]:
!python -c 'import SimpleHTTPServer; SimpleHTTPServer.test()' 8001

Serving HTTP on 0.0.0.0 port 8001 ...
127.0.0.1 - - [15/Apr/2015 23:21:32] "GET / HTTP/1.1" 200 -
^C----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 56218)
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 295, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 321, in process_request
    self.finish_request(request, client_address)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 649, in __init__
    self.handle()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/

- 디렉토리 리스트가 나오는 것은 SimpleHTTPRequestHandler의 do_GET() 메소드가 그렇게 구현되어 있기 때문

### 2.3.4 CGIHTTPServer 모듈

- do_POST() 메소드가 정의되어 있음
- 다만 CGIHTTPServer 클래스의 do_POST() 메소드는 CGI 처리 기능만 구현되어 있어서 모든 POST 방식을 처리할 수 있는 것은 아님

In [64]:
!python -m CGIHTTPServer 8001

Serving HTTP on 0.0.0.0 port 8001 ...
127.0.0.1 - - [15/Apr/2015 23:23:47] "GET / HTTP/1.1" 200 -
^CTraceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/CGIHTTPServer.py", line 377, in <module>
    test()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/CGIHTTPServer.py", line 373, in test
    SimpleHTTPServer.test(HandlerClass, ServerClass)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SimpleHTTPServer.py", line 220, in test
    BaseHTTPServer.test(HandlerClass, ServerClass)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/BaseHTTPSer

### 예제 2-14 CGIHTTPServer 시험용 클라이언트 - cgi_client.py

In [72]:
import urllib2
from urllib import urlencode

url = 'http://localhost:8001/cgi-bin/script.py'
data = {
    'language' : 'python',
    'framework': 'django',
    'email': 'kim@naver.com'
    }
encData = urlencode(data)

f = urllib2.urlopen(url, encData) # POST
print f.read()

Welcome, CGI Scripts
language is python
framework is django
email is kim@naver.com



### 2.3.5 xxxHTTPServer 모듈간의 관계

- 책에 그림으로 잘 설명되어 있음
- 상속 관계.

## 2.4 CGI/WSGI 라이브러리

- WSGI(Web Server Gateway Interface) 규격
- 전통적인 웹 CGI 기술의 단점을 개선하고 파이썬 언어에 맞게 재구성한 것
- 주의할 점은 CGI 기술은 이미 사라진 기술이지만, cgi 모듈의 라이브러리는 WSGI 기술에 여전히 사용되고 있다는 것


### 2.4.1 CGI 관련 모듈

- 웹 서버와 애플리케이션
- Static Request: 웹서버에 있는 파일을 있는 그대로 요청
- Dynamic Request: 현재의 시간을 요청하는 것처럼 동일한 요청이라도 시점에 따라 응답 내용이 달라짐
- 웹 서버에는 동적 요청을 애플리케이션에 넘겨주고 그 결과를 받는 기능이 필요함
- 이와 같이 웹 서버가 사용자의 요청을 애플리케이션에 전달하고 애플리케이션의 처리 결과를 애플리케이션으로부터 되돌려받기 위한, 즉 웹 서버와 애플리케이션 간의 데이터를 주고받기 위한 규격을 CGI(Common Gateway Interface)
- cgi 모듈: 요청에 포함된 파라미터를 처리하기 위한 FieldStorage 클래스 정의
- cgitb 모듈: CGI 애플리케이션(스크립트)이 실행 중 에러가 발생하는 경우, 에러에 대한 상세 정보를 브라우저에 표시
- WSGI 기술을 사용하여 CGI 처리를 함

### 2.4.2 WSGI 개요

- CGI 방식: 요청이 들어올 때마다 처리를 위한 프로세스를 생성하는 방식
- 웹 애플리케이션을 작성하는 개발자에게 이러한 방식을 선택하고 연동 방식을 맞추라고 한다면, 웹 애플리케이션 개발이 너무 어려워짐
- 그래서 웹 서버와 웹 애플리케이션 중간에서 이런 까다로운 처리를 해주는 것이 장고와 같은 웹 프레임워크
- 웹 애플리케이션을 한 번만 작성하면 다양한 웹 서버에서 동작
- 즉, 웹 서버에 독립적인 웹 애플리케이션을 작성할 수 있도록 웹 서버와 웹 애플리케이션 간에 연동규격을 정의한 것이 WSGI 규격

### 2.4.3 WSGI 서버의 애플리케이션 처리 과정

- WSGI 규격을 준수하는 WSGI 서버에서 실행되는 애플리케이션을 작성하는 것은 편리한 점이 많음
- 대부분의 파이썬 웹 프레임워크는 WSGI 서버를 제공하며, 애플리케이션 개발자는 WSGI 서버에 대한 API 규격만 맞추면, 웹 서버와는 독립적으로 애플리케이션을 작성할 수 있어 생산성이 높아짐

### WSGI 규격에 따라 애플리케이션을 개발할 때 중요한 사항 3가지

1. 개발이 필요한 애플리케이션을 함수 또는 클래스의 메소드로 정의하고, 애플리케이션 함수의 인자는 다음과 같이 정의
```python
def application_name(environ, start_response):
```

  - environ: 웹 프레임워크에 이미 정의되어 있으며, HTTP_HOST, HTTP_USER_AGENT, SERVER_PROTOCOL과 같은 HTTP 환경변수를 포함합니다
  - start_response: 애플리케이션 내에서 응답을 시작하기 위해 반드시 호출해야 하는 함수입니다
2. start_response 함수의 인자 역시 다음과 같이 정해져 있으므로 그대로 따릅니다.
```python
start_response(status, headers)
```

  - status: 응답 코드 및 응답 메세지를 지정합니다(200 OK, 404 Not Found 등)
  - headers: 응답 헤더를 포함합니다.
3. 애플리케이션 함수의 리턴값은 응답 바디에 해당하는 내용으로, 리스트나 제너레이터와 같은 iterable 타입이어야 합니다

### 2.4.4 wsgiref.simple_server 모듈

- 파이썬 표준 라이브러리에서는 웹 프레임워크 개발자가 웹 서버와의 연동 기능을 개발할 수 있도록 wsgiref 패키지의 하위 모듈로 wsgiref.simple_server 모듈을 제공하고 있음
- 이 모듈은 WSGI 스펙을 준수하는 웹 서버(일명, WSGI 서버)에 대한 참조(referecen) 서버, 즉 개발자에게 참고가 될 수 있도록 미리 만들어 놓은 WSGIServer 클래스와 WSGIRequestHandler 클래스를 정의하고 있습니다.
- 다만 모든 웹 프레임워크가 wsgiref 패키지를 사용하는 것은 아님
- wsgiref 패키지를 사용하지 않더라도, WSGI 스펙을 준수하는 자신만의 웹 프레임워크 또는 WSGI 서버를 만들면 되기 때문
- Flask 웹 프레임워크에서 사용하는 werkzeug WSGI 서버가 한 예로, wsgiref 패키지를 사용하지 않는 WSGI 웹 서버입니다.

### 예제 2-16 WSGI 서버

In [ ]:
def my_app(environ, start_response):
    status = '200 OK'
    headers = [('Content-Type', 'text/plain')]
    start_response(status, headers)
    
    return ['This is a sample WSGI Application.']

if __name__ == '__main__':
    from wsgiref.simple_server import make_server
    
    print 'Started WSGI Server on port 8001...'
    server = make_server('', 8001, my_app)
    server.serve_forever()

### WSGI 서버를 만들 때 중요한 사항

- wsgiref.simple_server 모듈은 WSGI 규격을 준수하여 WSGI 서버를 작성할 수 있도록 API를 제공하고 있으며, make_server() 및 serve_forever() 메소드가 그런 API의 일부
- my_app()과 같은 애플리케이션 로직을 호출 가능한(callable) 함수나 메소드로 정의하여, 이 함수를 make_server() 인자로 넘겨주어 WSGI 웹 서버를 만듭니다. 이는 애플리케이션 프로그램과 웹 서버 프로그램을 독립적으로 작성할 수 있게 해주는 WSGI 규격의 중요한 원칙에 해당함
- my_app() 함수가 WSGI 규격을 준수하는 애플리케이션 코드. 이 애플리케이션 함수에서 응답을 위한 헤더 및 바디를 구성해서 반환해줌. 헤더 및 바디를 구성하는 방법은 앞 절의 설명을 참고하기 바람

- WSGI 서버도 웹 서버이므로 WSGI 서버를 만드는 방식도 **2.3.1 간단한 웹 서버**에서 설명한 웹 서버를 만드는 방식을 그대로 따르고 있음
- 즉, 서버의 IP 및 PORT, 그리고 핸들러 클래스를 정의한 후에 이들을 인자로 하여 HTTPServer 객체를 생성합니다. 그런 다음 HTTPServer 객체의 serve_forever() 메소드를 호출하는 방식
- 다만 예제에서 만든 WSGI 서버에서 다른 점은 핸들러 클래스가 아니라 애플리케이션 로직을 작성하는 함수를 사용한다는 점과 애플리케이션 함수를 작성하는 방법입니다. 또한, 다음 그림처럼 HTTPServer 객체를 상속받은 WSGIServer 객체를 생성하여 serve_forever() 메소드를 호출한다는 점도 다릅니다. 물론 핸들러 클래스인 WSGIRequestHandler를 사용하여 WSGI 서버를 만들 수도 있습니다.